### PR TITLE
scan with pagination changes

### LINKED ISSUE #323: Format exception while running the wavi app with scan_with_pagination branch
**Describe the bug**
I tried to run the wavi app with the servers pointing to the scan_with_pagination branch. I see format exceptions while getting the server status

**To Reproduce**
1. First run the servers with the scan_with_pagination branch
2. Then install the wavi app and enter the atsign (servers running with scan_with_pagination) changes

**Expected behavior**
Server status should be returned. And atsign should be authenticated into the app

**Screenshots**

**Were you using an @‎application when the bug was found?**
  - [e.g. @‎ wavi]

**Additional context**
**App logs**

**For already authenticated atsigns**

FINER|2021-09-28 14:42:47.055861|AtLookup|AtLookup.findSecondary got answer: vip.ve.atsign.zone:4001 and closing connection
I/flutter (12045): _getServerStatus error: FormatException: Unexpected character (at character 20)
I/flutter (12045): "publickey@ sitaram","signing_publickey@ sitaram"
I/flutter (12045): ^
I/flutter (12045): outbound finish handler called

**For a new atsign**

: FINER|2021-09-28 14:42:03.585260|AtLookup|AtLookup.findSecondary got answer: vip.ve.atsign.zone:6365 and closing connection
I/flutter (12045): _getServerStatus error: type 'String' is not a subtype of type 'Iterable<dynamic>'
I/flutter (12045): outbound finish handler called

Server Logs
Screenshot from 2021-09-28 16-35-25

### PR DESCRIPTION
**Pull request by naresh0689**
**- What I did**
Added pagination functionality to scan verb. 
**- How I did it**
If we mention pageId, scan verb will return keys in the page along with total pages, keys per page. 
**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: at_secondary/at_secondary_server/lib/src/verb/handler/scan_verb_handler.dart [hunk 1]
```diff
@@ -46,7 +46,8 @@ class ScanVerbHandler extends AbstractVerbHandler {
         atConnection.getMetaData() as InboundConnectionMetadata;
     var forAtSign = verbParams[FOR_AT_SIGN];
     var scanRegex = verbParams[AT_REGEX];
-
+    var pageId = verbParams['pageId'];
+    print('pageId : $pageId');
```
Comment (murali-shris) [inline at_secondary/at_secondary_server/lib/src/verb/handler/scan_verb_handler.dart]: please remove print statement

File: at_secondary/at_secondary_server/lib/src/verb/handler/scan_verb_handler.dart [hunk 2]
```diff
@@ -60,12 +60,19 @@ class ScanVerbHandler extends AbstractVerbHandler {
       if (forAtSign != null &&
           atConnectionMetadata.isAuthenticated &&
           forAtSign != currentAtSign) {
-        response.data =
-            await _getExternalKeys(forAtSign, scanRegex, atConnection);
+        if (page == null) {
```
Comment (VJag) [inline at_secondary/at_secondary_server/lib/src/verb/handler/scan_verb_handler.dart:63]: Even when the keys are external, that secondary's can should know how to give back results for a page. We should not get all of the results and then return a sublist.

File: at_secondary/at_secondary_server/lib/src/verb/handler/scan_verb_handler.dart [hunk 3]
```diff
@@ -60,12 +60,19 @@ class ScanVerbHandler extends AbstractVerbHandler {
...
+          response.data =
+              await _getExternalKeys(forAtSign, scanRegex, atConnection);
+        } else {
+          var keysString =
+              await _getExternalKeys(forAtSign, scanRegex, atConnection);
+          response.data =
+              _prepareExternalKeysResponse(int.parse(page), keysString);
+        }
+        print('external keys : ${response.data}');
```
Comment (VJag) [inline at_secondary/at_secondary_server/lib/src/verb/handler/scan_verb_handler.dart]: Please remove prints and add appropriate logs. Let's not log data as it is open of the main reason for the memory spike on the secondaries.

File: at_secondary/at_secondary_server/lib/src/verb/handler/scan_verb_handler.dart [hunk 4]
```diff
@@ -133,17 +149,66 @@ class ScanVerbHandler extends AbstractVerbHandler {
...
+      result['keys'] = jsonEncode([]);
+    }
+    result['totalPages'] =
+        ((keys.length) / 10).toInt() + ((keys.length) % 10 > 0 ? 1 : 0);
+    result['keysPerPage'] = 10;
+    result['currentPage'] = page;
+    return jsonEncode(result);
+  }
+
+  String? _prepareLocalKeysResponse(int page, List<dynamic>? keys) {
```
Comment (VJag) [inline at_secondary/at_secondary_server/lib/src/verb/handler/scan_verb_handler.dart:171]: _prepareLocalKeysResponse and _prepareExternalKeysResponse have similar logic. This needs to be refactored to a single method called _preparePaginatedResponse and reuse.

File: at_secondary/at_secondary_server/lib/src/verb/handler/scan_verb_handler.dart [hunk 5]
```diff
@@ -60,12 +60,18 @@ class ScanVerbHandler extends AbstractVerbHandler {
       if (forAtSign != null &&
           atConnectionMetadata.isAuthenticated &&
           forAtSign != currentAtSign) {
-        response.data =
-            await _getExternalKeys(forAtSign, scanRegex, atConnection);
+        if (page == null) {
+          response.data =
```
Comment (VJag) [inline at_secondary/at_secondary_server/lib/src/verb/handler/scan_verb_handler.dart:64]: I don't think we need a if-else here. When we ask an external secondary for a scan with the page its response should be ready to be sent back without any work. We should be sending the page based syntax to it.

File: at_secondary/at_secondary_server/lib/src/verb/handler/scan_verb_handler.dart [hunk 6]
```diff
@@ -77,7 +83,12 @@ class ScanVerbHandler extends AbstractVerbHandler {
             ? response.data?.split(',')
```
Comment (VJag) [inline at_secondary/at_secondary_server/lib/src/verb/handler/scan_verb_handler.dart:83]: logger.finer('response.data : ${response.data}'); Lets not have it in line 81

Comment (VJag) [reply at_secondary/at_secondary_server/lib/src/verb/handler/scan_verb_handler.dart:83]: Or may be have it with least level of logging and ideally servers should run with debug level

**Comment by murali-shris** [comment]
closing this PR since it is obsolete

**Pull request closed without merging**